src/autopas/iterators/RegionParticleIterator.h

```diff
@@ -25,50 +27,101 @@ class RegionParticleIterator : public ParticleIterator<Particle, ParticleCell> {
+  explicit RegionParticleIterator(std::vector<ParticleCell> *cont, const std::array<size_t, 3> cellsPerDimension,
+                                  std::array<double, 3> startRegion, std::array<double, 3> endRegion, size_t startIndex,
+                                  size_t endIndex, CellBorderAndFlagManager *flagManager = nullptr,
                                   IteratorBehavior behavior = haloAndOwned)
-      : ParticleIterator<Particle, ParticleCell>(cont, flagManager, behavior),
+      : ParticleIterator<Particle, ParticleCell>(cont, startIndex, flagManager, behavior),
+        _cellsPerDim(cellsPerDimension),
         _startRegion(startRegion),
-        _endRegion(endRegion) {
+        _endRegion(endRegion),
+        _startIndex(startIndex),
+        _endIndex(endIndex) {
+    _startIndex3D = utils::ThreeDimensionalMapping::oneToThreeD(_startIndex, _cellsPerDim);
+    _endIndex3D = utils::ThreeDimensionalMapping::oneToThreeD(_endIndex, _cellsPerDim);
+
+    _shortJump = _cellsPerDim[0] - (_endIndex3D[0] - _startIndex3D[0]) - 1;
+    _longJump = _cellsPerDim[0] * (_cellsPerDim[1] - (_endIndex3D[1] - _startIndex3D[1]) - 1);
+
     // ParticleIterator's constructor will initialize the Iterator, such that it
     // points to the first particle if one is found, otherwise the pointer is
     // not valid
-    if (this->isValid()) {  // if there is NO particle, we can not dereference it, so we need a check.
-      if (not(this->operator*()).inBox(_startRegion, _endRegion)) {
+    if (ParticleIterator<Particle, ParticleCell>::isValid()) {  // if there is NO particle, we can not dereference it,
+                                                                // so we need a check.
+      if (notInBox(this->operator*().getR(), _startRegion, _endRegion)) {
         operator++();
       }
     }
   }
 
   /**
    * @copydoc ParticleIteratorInterface::operator++
-   * @todo optimize! this version is currently very slow
    */
   inline RegionParticleIterator<Particle, ParticleCell> &operator++() override {
     do {
       ParticleIterator<Particle, ParticleCell>::operator++();
     } while (ParticleIterator<Particle, ParticleCell>::isValid() &&
-             !(this->operator*()).inBox(_startRegion, _endRegion));
+             notInBox(this->operator*().getR(), _startRegion, _endRegion) && this->getCurrentCellId() <= _endIndex);
     return *this;
   }
 
+  bool isValid() const override {
+    return ParticleIterator<Particle, ParticleCell>::isValid() &&
+           inBox(this->operator*().getR(), _startRegion, _endRegion);
+  }
+
   // @todo add test of clone
   inline ParticleIteratorInterfaceImpl<Particle> *clone() const override {
     return new RegionParticleIterator<Particle, ParticleCell>(*this);
   }
 
  private:
+  /**
+   * @copydoc
+   * @note overrides call in ParticleIterator<Particle, ParticleCell>::operator++()
+   */
+  void next_non_empty_cell() override {
```

Comment:
slower -> probably since more calculations for index conversions are needed.
easier to read -> not sure about that since the jumps are pretty wild :D